components/dashboard/src/settings/EnvironmentVariables.tsx

```diff
@@ -4,12 +4,195 @@
+    const [isAddEnvVarModalVisible, setAddEnvVarModalVisible] = useState(false);
+    const update = async () => {
+        await getGitpodService().server.getEnvVars().then(r => setEnvVars(r));
+    }
+
+    useEffect(() => {
+        update()
+    }, []);
+
+
+    const add = () => {
+        setCurrentEnvVar({ name: '', value: '', repositoryPattern: '' });
+        setAddEnvVarModalVisible(true);
+    }
+
+    const edit = (ev: UserEnvVarValue) => {
+        setCurrentEnvVar(ev);
+        setAddEnvVarModalVisible(true);
+    }
+
+    const save = async (variable: UserEnvVarValue) => {
+        await getGitpodService().server.setEnvVar(variable);
+        await update();
+    };
+
+    const deleteV = async (variable: UserEnvVarValue) => {
+        await getGitpodService().server.deleteEnvVar(variable);
+        await update();
+    };
+
+    const validate = (variable: UserEnvVarValue) => {
+        const name = variable.name;
+        const pattern = variable.repositoryPattern;
+        if (name.trim() === '') {
+            return 'Name must not be empty.';
+        }
+        if (!/^[a-zA-Z0-9_]*$/.test(name)) {
+            return 'Name must match /[a-zA-Z_]+[a-zA-Z0-9_]*/.';
+        }
+        if (variable.value.trim() === '') {
+            return 'Value must not be empty.';
+        }
+        if (pattern.trim() === '') {
+            return 'Scope must not be empty.';
+        }
+        const split = pattern.split('/');
+        if (split.length < 2) {
+            return "A scope must use the form 'organization/repo'.";
+        }
+        for (const name of split) {
+            if (name !== '*') {
+                if (!/^[a-zA-Z0-9_\-.\*]+$/.test(name)) {
+                    return 'Invalid scope segment. Only ASCII characters, numbers, -, _, . or * are allowed.';
+                }
+            }
+        }
+        return '';
+    };
+
+    return <SettingsPage title='Variables' subtitle='Configure environment variables for all workspaces.'>
+        {isAddEnvVarModalVisible ? <AddEnvVarModal
+            save={save}
+            envVar={currentEnvVar}
+            validate={validate}
+            onClose={() => setAddEnvVarModalVisible(false)} /> : null}
+        {envVars.length === 0
+            ? <div className="bg-gray-100 rounded-xl w-full h-96">
+                <div className="pt-28 flex flex-col items-center w-96 m-auto">
+                    <h3 className="text-center pb-3">No Environment Variables</h3>
+                    <div className="text-center pb-6 text-gray-500">In addition to user-specific environment variables you can also pass variables through a workspace creation URL. <a className="text-gray-400 underline underline-thickness-thin underline-offset-small hover:text-gray-600" href="https://www.gitpod.io/docs/env-vars/">Learn more</a></div>
+                    <button onClick={add} className="font-medium">New Environment Variable</button>
+                </div>
+            </div>
+            : <div>
```

Comment:
**issue**: Spacing missing.

**thought**: The amount of semantic inconsistencies (HTML structure) between the same elements and components is slightly increasing! ⚠️ 

```suggestion
            : <div class="space-y-2">
```

Reply:
I've applied the change as suggested but wonder if it is not better to apply space on the childs or at least only on one or the other not on both.

Reply:
Absolutely, it would be better to apply spacing on the child elements as the filter section could be a separate component in the long run (see design system).

The suggestion was due to existing usage[[1](https://github.com/gitpod-io/gitpod/blob/8920df5c7da91a37157cce476753f6c99386bd66/components/dashboard/src/workspaces/Workspaces.tsx#L100)][[2](https://github.com/gitpod-io/gitpod/blob/df8b5cf582a8e0818b094858c6c196c8c2af9ac2/components/dashboard/src/settings/Integrations.tsx#L242)][[3](https://github.com/gitpod-io/gitpod/blob/864b83c8ec9e368ad7f41d6a49e8736bb60063eb/components/dashboard/src/workspaces/Workspaces.tsx#L100)][...] in order to keep the same structure (technical debt).